File: vehicle/haskell/src/Hybrid.hs
module Hybrid where

import Aircraft as A (Aircraft (..))
import Car as C (Car (..))
import Submarine as S (Submarine (..))
import Vehicle as V (Vehicle (..))

data Hybrid = Hybrid

instance Car Hybrid

instance Aircraft Hybrid

instance Submarine Hybrid

instance Vehicle Hybrid where
  getEngine _ = "hybrid engine"

  simulate hybrid = do
    putStrLn ("Vehicle engine: " <> V.getEngine hybrid)

    C.drive hybrid
    A.pilot hybrid
    S.dive hybrid

File: vehicle/rust/src/hybrid.rs
use crate::{Aircraft, Car, Submarine, Vehicle};

pub struct Hybrid {}

impl Hybrid {
  pub fn new() -> Hybrid {
    Hybrid {}
  }
}

impl Car for Hybrid {}

impl Aircraft for Hybrid {}

impl Submarine for Hybrid {}

impl Vehicle for Hybrid {
  fn get_engine(&self) -> String {
    String::from("hybrid engine")
  }

  fn simulate(&self) {
    println!("Vehicle engine: {}", Vehicle::get_engine(self));

    Car::drive(self);
    Aircraft::pilot(self);
    Submarine::dive(self);
  }
}
